### PR TITLE
storage: introduce healthchecker

### PR DESCRIPTION
**Pull request by andrioni**
The Healthchecker is a new component of the source ingestion pipeline which keeps track of the health of a given source.

This PR is the follow-up to #13753, adding the code of the Healthchecker and including support for ingesting new updates from sources, but without actually changing any sources to push status updates, which will come in a future PR.

### Motivation
  * This PR makes progress towards a known-desirable feature: #12864

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:321]: Do you have ideas about when a source would ever be in this state? I naively thought users could always keep querying what's already there in persist.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:51]: This is `SourceData` because that's what's in all the other shards that are backing sources or tables, right?

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:169]: nit: these log messages seem hard to understand without context.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:190]: Same as above, I _feel_ like we could retry on this one if we sync up, like for the `Ok(Ok(Err` case.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:787]: there is `read_all()` on `SnapshotIter` but it's behind `cfg(test)` and it seems that can't be used from tests of other modules. :see_no_evil:

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:796]: nit: should be sth like `get_upper` or `get_collection_upper`

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:709]: @danhhz What do you think about these helpers for testing? Is that how we want other modules to test things that integrate with persist? Should we provide helpers with persist so that we don't have do duplicate these in the future?

**Review by aljoscha — CHANGES_REQUESTED** [review]
This looks very neat! :tada: I had some nits/questions/comments. And it might be good to wait for Dan's feedback on how the persist/testing integration looks.

The only thing that I would like to see before merging is to resolve the error handling/retires at the two `compare_and_append` callsites.

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:787]: Yep, I did try using it, and this is an unashamed copy of that exactly because it didn't work :(

File: src/storage/src/source/mod.rs
Comment (andrioni) [reply src/storage/src/source/mod.rs:750]: Just to make it explicit, right? Because the runtime behavior is the same, as far I understand (I think I might have asked you about this some time ago)

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:29]: It is not! It is mostly to make it clear that this is instantiated per worker (in my understanding of the code), instead of per process or similar.

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:37]: Mostly because all the times this is going to be used later (building Rows and matching against Rows) it is going to compare against the string serialization of it, so I wanted to pay the conversion cost once. I can switch it to `GlobalId` if you think it's better, though!

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:54]: oops, nice catch! The docstring was indeed ~shamelessly copied~ borrowed from the reclocking operator.

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:89]: Because the `snapshot` happens at the `since`, as it can't happen at the `upper`, and doing `upper - 1` felt a bit too much like hard-coding that our timestamps are `u64`s. And AFAIU, we could lose updates if the `listener` was started at `upper` if the snapshot was generated at `upper`.

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:143]: Ah, nice catch, the `loop` is there mostly because this was originally ~copy-and-pasted~ inspired by `update_status`. I also didn't pay enough attention to what this specific error meant, in my mind it was more of a "things are so out of control that it's better to panic and restart storaged" than "oh, S3 is just having a bad day and we got timed out".

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:321]: They can't if we ever push something to error collection of a source! That said, the most common case is for decoding errors, which are unfortunately not captured here anyway, as this happens further down in the processing pipeline, but I was thinking we could potentially do something there later on.

I think [this Postgres source codepath](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/postgres.rs#L325-L328) is the main relevant way this can happen right now, but I haven't reproduced it locally yet.

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:51]: Exactly! Should I add a comment about it? This is how all our tables (and I believe sources as well) are persisted.

File: src/storage/src/source/mod.rs
Comment (aljoscha) [reply src/storage/src/source/mod.rs:750]: Yes, and (I think) yes :sweat_smile:. Plus, I'm quite sure we have a lint that disallows your code.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [reply src/storage/src/source/healthcheck.rs:37]: eh, no strong opinion. I just like strongly typed things and we could keep that "information" as long as possible.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [reply src/storage/src/source/healthcheck.rs:51]: probably would be good, yes

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [reply src/storage/src/source/healthcheck.rs:143]: These errors can happen in situations like "your append was succesful but sth timed out and we didn't hear back so we don't know for sure if it was successful or not". And in cases where it's not harmful to try again and potentially advance the timestamp a bit further. Or in cases we sync up our state and then see if we still need to emit our status update it's good to retry.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [reply src/storage/src/source/healthcheck.rs:89]: yeah, this is indeed a bit rough. Let's keep it like this, and maybe add what you commented here as a comment, wdyt?

Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:89]: Sounds good to me!

File: src/storage/src/source/healthcheck.rs
Comment (andrioni) [reply src/storage/src/source/healthcheck.rs:709]: I copied most of this logic from the reclocking tests (with some snippets coming from other persist-related tests), so definite +1 from me on finding ways to not avoid all this copy-and-pasting (although my Rust-fu is not good enough to propose a way to make it better, as sharing test code between modules proved harder than I expected).

Comment (danhhz) [reply src/storage/src/source/healthcheck.rs:709]: My opinion is that unit tests should create one PeristClientCache/PersistClient at the top of the test and plumb it wherever needed. This lazy static thing was something petros wanted so he could share state (which did and still feels like a testing anti-pattern to me) 🤷

File: src/storage/src/source/healthcheck.rs
Comment (danhhz) [reply src/storage/src/source/healthcheck.rs:787]: it might work if you make that `cfg(any(test, debug_assertions))`?

File: src/storage/src/source/mod.rs
Comment (danhhz) [reply src/storage/src/source/mod.rs:750]: I know we have that lint in persist, but did it ever land repo wide?

**Comment by andrioni** [comment]
After a rebase with `main`, something in this branch is causing a very weird panic with an unrelated relation, currently looking into it.

File: src/storage/src/source/healthcheck.rs
Comment (aljoscha) [inline src/storage/src/source/healthcheck.rs:160]: :raised_hands: Good, concise explanation!

**Review by aljoscha — APPROVED** [review]
:+1:

**Pull request merged**
